### PR TITLE
JS: Decompression Bombs

### PR DESCRIPTION
**Pull request by am0o0**
This is part of All for one, one for all query submission, I'm going to submit an issue in github/securitylab for this pull request too.
I've added sanitizers as much as I could find safe ways of using methods of each library in their documentation.
I also added some additional taint steps which I think it is good to consider them as global steps too.
I tried my best to cover Command line flow sources as much as possible which can be used in other existing query too.
One interesting thing is that during finding CVEs for this submission I found out that there are some second level Remote flow sources like a situation that users upload their files and then get the file ID in one endpoint and their web app client will send another request containing this file ID to a second endpoint. the second endpoint mostly need a additional taint steps to work, one of these steps is Sequelize library Model. these additional steps are like second level Remote flow sources which is come after first level ones.

### CODE REVIEW

**Comment by smowton** [comment]
Anyone not watching the repo in general, note this is part of a family of submissions:

https://github.com/github/codeql/pull/13553
https://github.com/github/codeql/pull/13554
https://github.com/github/codeql/pull/13555
https://github.com/github/codeql/pull/13556
https://github.com/github/codeql/pull/13557
https://github.com/github/codeql/pull/13558

**Review by asgerf — COMMENTED** [review]
Hi @amammad 👋 

Thanks for the many submissions. I haven't had a chance to review the impact, but it certainly seems very valuable to have these queries.

Before we get into language-specific reviews, there are a few recurring technical issues with the PRs:
- There seem to be some accidental file deletions/modifications of random unrelated files in the repo. Please undo those changes.
- Ensure the `.qhelp` file has the same name as the corresponding `.ql` file. Otherwise it just doesn't work and the PR can't be merged.
- Stick to one query per language.
  - The JS PR contains one query per library. They will need to be merged into one query.
  - The Ruby PR has two variants of the query. Please pick one of them to keep.

If you could address the above for all the PRs that would be help the review process a lot.

Lastly, once the above has been addressed, it would be great if you could mention when you think the PR is ready for review, at which point you should stop pushing new changes, unless in response to a review comment. It makes reviews very hard when new changes keep coming in.

File: javascript/ql/lib/semmle/javascript/frameworks/FormParsers.qll [hunk 1]
```diff
@@ -3,61 +3,188 @@
...
-    )
+    }
   }
+}
 
-  override string getSourceType() { result = "parsed user value from Multiparty" }
+/**
+ * An Additional taint step like `for (succ in pred)`
+ */
+private class AdditionalTaintStepForIn extends TaintTracking::SharedTaintStep {
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/FormParsers.qll:184]: I don't like this as a general taint step.  
The iteration variable in a `for (x in y)` statement holds the keys of the iteration domain, and not the values.  
It's way to general to use as a taint-step in all our queries.  

But I can see that the header names are sometimes tainted.  
I think the proper solution is to use flow-labels, which can help you track different kinds of values at the same time.  
And then add the flow steps to a `isAdditionalTaintStep` in your `Configuration` class.

You can try to look at `Xss.ql` or `TaintedPath.ql` for examples of how flow-labels can be used.  

In this case you want your header sources to start with 2 flow labels (`taint`, and `header-object`).  
And then have an extra taint step that converts the label from `header-object` to `taint` if the keys of the object are read (e.g. in a `for in` statement).  

---- 

But.  
This PR is very complicated as it is, and would take a lot of work to get merged.
If you want any change of getting this merged, then I suggest you move it to the `experimental/` folder.

[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/FormParsers.qll]: It's simpler if you move all the code to `experimental/`.

**Pull request merged**